### PR TITLE
Added support for forcing files to be zipped when uploading

### PR DESCRIPTION
**Pull request by dtimberlake**
When packaging lambda functions with `aws cloudformation package`, the code must be zipped even if it's a single file. This is in contrast to other resources (i.e. stack templates or api swagger files). The error given by CloudFormation when deploying a template specifying a single code file (after running the package command) is `Could not unzip uploaded file. Please check your file, then try to upload again.`

To accommodate this difference, I've added support for a `FORCE_ZIP` attribute on a `Resource` to determine if it's property value should always be zipped. It accomplishes this by copying the file into a temporary directory, zipping the directory, uploading the zipped file, then discarding the temporary directory.

I have added the new `FORCE_ZIP` attribute to `LambdaFunctionResource` and `ServerlessFunctionResource`.

### CODE REVIEW

**Comment by jamesls** [comment]
Thanks for the pull request, we'll take a look.

Looping in the service team:
cc @sanathkr

**Comment by Sakshi99** [comment]
hi Jamesls ? how are you doing?

**Comment by sanathkr** [comment]
Can you make sure shutil.* works on Python 2.6? Otherwise this looks good to me

**Comment by dtimberlake** [comment]
I tested `shutil` in 2.6.9 and it worked fine. It was already part of the standard lib by that point.

**Comment by dtimberlake** [comment]
Any movement on this?

File: awscli/customizations/cloudformation/artifact_exporter.py [hunk 1]
```diff
@@ -199,12 +200,20 @@ def mktempfile():
...
+    return tmp_dir
+ 
+
 class Resource(object):
     """
     Base class representing a CloudFormation resource that can be exported
     """
 
     PROPERTY_NAME = None
+    FORCE_ZIP = False
```
Comment (pmarques) [inline awscli/customizations/cloudformation/artifact_exporter.py]: I think this should be true by default, otherwise the fix doesn't work by default.

**Comment by pmarques** [comment]
Instated of moving files around why not keep it even more simple? This is my current patch:

```
commit ffb80d6412cc1c83bcd0adf923bf867255628438 (HEAD -> develop)
Author: Patrick F. Marques <patrick.marques@automationlogic.com>
Date:   Thu Jul 20 16:48:23 2017 +0100

    Force Lambda function code inside a zip @S3

diff --git a/awscli/customizations/cloudformation/artifact_exporter.py b/awscli/customizations/cloudformation/artifact_exporter.py
index 105557ef..46fa1eba 100644
--- a/awscli/customizations/cloudformation/artifact_exporter.py
+++ b/awscli/customizations/cloudformation/artifact_exporter.py
@@ -88,7 +88,7 @@ def parse_s3_url(url,
 
 
 def upload_local_artifacts(resource_id, resource_dict, property_name,
-                           parent_dir, uploader):
+                           parent_dir, uploader, force_zip=False):
     """
     Upload local artifacts referenced by the property at given resource and
     return S3 URL of the uploaded object. It is the responsibility of callers
@@ -131,7 +131,7 @@ def upload_local_artifacts(resource_id, resource_dict, property_name,
     local_path = make_abs_path(parent_dir, local_path)
 
     # Or, pointing to a folder. Zip the folder and upload
-    if is_local_folder(local_path):
+    if is_local_folder(local_path) or force_zip:
         return zip_and_upload(local_path, uploader)
 
     # Path could be pointing to a file. Upload the file
@@ -268,7 +268,8 @@ class ResourceWithS3UrlDict(Resource):
         artifact_s3_url = \
             upload_local_artifacts(resource_id, resource_dict,
                                    self.PROPERTY_NAME,
-                                   parent_dir, self.uploader)
+                                   parent_dir, self.uploader,
+                                   self.ALWAYS_ZIP)
 
         resource_dict[self.PROPERTY_NAME] = parse_s3_url(
                 artifact_s3_url,
@@ -292,6 +293,7 @@ class LambdaFunctionResource(ResourceWithS3UrlDict):
     BUCKET_NAME_PROPERTY = "S3Bucket"
     OBJECT_KEY_PROPERTY = "S3Key"
     VERSION_PROPERTY = "S3ObjectVersion"
+    ALWAYS_ZIP = true
 
 
 class ApiGatewayRestApiResource(ResourceWithS3UrlDict):
```

**Comment by pmarques** [comment]
I created a patch running on my deploy environments that is not using the module shutil:

```
--- awscli/customizations/cloudformation/artifact_exporter.py
+++ awscli/customizations/cloudformation/artifact_exporter.py
@@ -88,7 +88,7 @@ def parse_s3_url(url,
 
 
 def upload_local_artifacts(resource_id, resource_dict, property_name,
-                           parent_dir, uploader):
+                           parent_dir, uploader, force_zip=False):
     """
     Upload local artifacts referenced by the property at given resource and
     return S3 URL of the uploaded object. It is the responsibility of callers
@@ -131,7 +131,7 @@ def upload_local_artifacts(resource_id, resource_dict, property_name,
     local_path = make_abs_path(parent_dir, local_path)
 
     # Or, pointing to a folder. Zip the folder and upload
-    if is_local_folder(local_path):
+    if is_local_folder(local_path) or force_zip:
         return zip_and_upload(local_path, uploader)
 
     # Path could be pointing to a file. Upload the file
@@ -176,6 +176,8 @@ def make_zip(filename, source_root):
     with open(zipfile_name, 'wb') as f:
         zip_file = zipfile.ZipFile(f, 'w', zipfile.ZIP_DEFLATED)
         with contextlib.closing(zip_file) as zf:
+            if os.path.isfile(source_root):
+                zf.write(source_root, os.path.basename(source_root))
             for root, dirs, files in os.walk(source_root):
                 for filename in files:
                     full_path = os.path.join(root, filename)
@@ -268,7 +270,8 @@ class ResourceWithS3UrlDict(Resource):
         artifact_s3_url = \
             upload_local_artifacts(resource_id, resource_dict,
                                    self.PROPERTY_NAME,
-                                   parent_dir, self.uploader)
+                                   parent_dir, self.uploader,
+                                   self.ALWAYS_ZIP)
 
         resource_dict[self.PROPERTY_NAME] = parse_s3_url(
                 artifact_s3_url,
@@ -292,6 +295,7 @@ class LambdaFunctionResource(ResourceWithS3UrlDict):
     BUCKET_NAME_PROPERTY = "S3Bucket"
     OBJECT_KEY_PROPERTY = "S3Key"
     VERSION_PROPERTY = "S3ObjectVersion"
+    ALWAYS_ZIP = True
 
 
 class ApiGatewayRestApiResource(ResourceWithS3UrlDict):
```

**Comment by pmarques** [comment]
@sanathkr @jamesls this PR seems to be forgotten....

**Comment by sanathkr** [comment]
Looks good to me. @dtimberlake Can set FORCE_ZIP=True by default and push a new revision? 

@JordonPhillips Feel free to merge once a new PR is submitted that is ready to be merged

**Comment by sanathkr** [comment]
Taking over this pull request in #2866 by adding more unit tests and covering the corner case where the local file could already be a zipfile. I'd like someone to review and take this in.

**Comment by jamesls** [comment]
Merged via https://github.com/aws/aws-cli/pull/2866
